oauth2/hook.go

```diff
@@ -66,103 +70,159 @@ type RefreshTokenHookResponse struct {
-			Session:         session,
-			Requester:       requesterInfo,
-			Subject:         session.GetSubject(),
-			ClientID:        requester.GetClient().GetID(),
-			GrantedScopes:   requester.GetGrantedScopes(),
-			GrantedAudience: requester.GetGrantedAudience(),
-		}
-		reqBodyBytes, err := json.Marshal(&reqBody)
-		if err != nil {
-			return errorsx.WithStack(
-				fosite.ErrServerError.
-					WithWrap(err).
-					WithDescription("An error occurred while encoding the refresh token hook.").
-					WithDebugf("Unable to encode the refresh token hook body: %s", err),
-			)
-		}
+func callHook(ctx context.Context, reg x.HTTPClientProvider, requester fosite.AccessRequester, hookType string, hookURL *url.URL) error {
+	if !requester.GetGrantTypes().ExactOne(hookType) {
+		return nil
+	}
 
-		req, err := retryablehttp.NewRequestWithContext(ctx, http.MethodPost, hookURL.String(), bytes.NewReader(reqBodyBytes))
-		if err != nil {
-			return errorsx.WithStack(
-				fosite.ErrServerError.
-					WithWrap(err).
-					WithDescription("An error occurred while preparing the refresh token hook.").
-					WithDebugf("Unable to prepare the HTTP Request: %s", err),
-			)
-		}
-		req.Header.Set("Content-Type", "application/json; charset=UTF-8")
-
-		resp, err := reg.HTTPClient(ctx).Do(req)
-		if err != nil {
-			return errorsx.WithStack(
-				fosite.ErrServerError.
-					WithWrap(err).
-					WithDescription("An error occurred while executing the refresh token hook.").
-					WithDebugf("Unable to execute HTTP Request: %s", err),
-			)
-		}
-		defer resp.Body.Close()
+	session, ok := requester.GetSession().(*Session)
+	if !ok {
+		return nil
+	}
 
-		switch resp.StatusCode {
-		case http.StatusOK:
-			// Token refresh permitted with new session data
-		case http.StatusNoContent:
-			// Token refresh is permitted without overriding session data
-			return nil
-		case http.StatusForbidden:
-			return errorsx.WithStack(
-				fosite.ErrAccessDenied.
-					WithDescription("The refresh token hook target responded with an error.").
-					WithDebugf("Refresh token hook responded with HTTP status code: %s", resp.Status),
-			)
-		default:
-			return errorsx.WithStack(
-				fosite.ErrServerError.
-					WithDescription("The refresh token hook target responded with an error.").
-					WithDebugf("Refresh token hook responded with HTTP status code: %s", resp.Status),
-			)
-		}
+	payload := map[string][]string{}
 
-		var respBody RefreshTokenHookResponse
-		if err := json.NewDecoder(resp.Body).Decode(&respBody); err != nil {
-			return errorsx.WithStack(
-				fosite.ErrServerError.
-					WithWrap(err).
-					WithDescription("The refresh token hook target responded with an error.").
-					WithDebugf("Response from refresh token hook could not be decoded: %s", err),
-			)
-		}
+	if requester.GetGrantTypes().ExactOne("urn:ietf:params:oauth:grant-type:jwt-bearer") || requester.GetGrantTypes().ExactOne("client_credentials") {
+		payload = requester.GetRequestForm()
```

Comment:
Added request payload to hook request only for jwt_bearer and client_credentials grant types. Other grant types don't have anything in their payload that could be useful in the hook.

Reply:
Makes sense!